### PR TITLE
[FluidDynamics] Extend DragUtilities

### PR DESCRIPTION
**Pull request by matekelemen**
**Description**
Please note that this is just a draft for now.

Add functionality to ```DragUtilities``` that allows to sum up more than one quantity when looping through nodes/elements. For example, this allows to compute the total body fitted drag and its moment on a point at the same time.

**Changelog**
- ```AtomicAdd``` for arrays extended to any array length (is there a reason why it was restricted to length 3 so far?)
- generalized functions added to ```DragUtilities``` that perform a set of operations on nodes/elements and sum up their results
- existing drag computing functions (eg. ```DragUtilities::CalculateBodyFittedDrag```) use the new generalized functions and parallel utilities instead of hard-coded OpenMP loops 
- none of the existing interfaces/signatures have not been changed

### CODE REVIEW

**Comment by matekelemen** [comment]
I can't find anything that would actually run the cpp tests in ```FluidDynamics``` so I couldn't test my changes yet. Additionally, I'd need a reasonably big test model with pre-computed reactions to check the performance, as requested by @mpentek. Can you send me one?

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
Several concerns. Main ones:
- I think that the same behavior can be achieved with way less complexity (assuming that the intention is to compute both the drag and the moment).
- Changes in the `atomic_utilities.h` must be discussed in a standalone PR. If I'm not wrong, the method was intentionally overloaded for `array_1d<double,3>` type to ease its usage without specifying the type. Aside of this, this is the type used in Kratos vector variables.

**Comment by mpentek** [comment]
> I can't find anything that would actually run the cpp tests in `FluidDynamics` so I couldn't test my changes yet. Additionally, I'd need a reasonably big test model with pre-computed reactions to check the performance, as requested by @mpentek. Can you send me one?

@matekelemen the testing we can discuss separately, as it is not strictly relevant for this PR

**Comment by mpentek** [comment]
> Several concerns. Main ones:
> 
> * I think that the same behavior can be achieved with way less complexity (assuming that the intention is to compute both the drag and the moment).
> * Changes in the `atomic_utilities.h` must be discussed in a standalone PR. If I'm not wrong, the method was intentionally overloaded for `array_1d<double,3>` type to ease its usage without specifying the type. Aside of this, this is the type used in Kratos vector variables.

@rubenzorrilla 

There are multiple aspects:

- the first would be the extend the drag calculation for moments as well, while keeping the existing function call possible
- additoinal would be to remove duplicate or similar functionalities (see [ExaquteSandboxApplication ->
- custom_utilities -> drag_and_moment_utilities](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/ExaquteSandboxApplication/custom_utilities))
- additional aspect would be to make the formulation more general as a sort of "GETter" function, such that other applications can benefit from it ( such as the StructuramMechanicsApplication -> to get REACTION and MOMENT_REACTION reduced to a point)
- if we have a getter, one could have the inverse opretation (for example [MeshMovingApplication
- custom_utilities -> move_mesh_utilities](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/MeshMovingApplication/custom_utilities) the could benefit from it for setting a rigid-body type translation + rotation of a model part, useful for FSI)

The first two of those strictly related to functionalities of the FluidDynamicsApplication. It is a question whether already for this PR it is worth to work even on a more generic fomulation to accomocate similiar later formulations. Or do you @rubenzorrilla  think that we should rather think on the application level?

As far as making the fomrulation generic and potentially enabling other applications to benefit from it, should we think about something in line with what can be seen in [kratos -> utilities -> variable_redistribution_utility.h](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/variable_redistribution_utility.h)? Also make additions/changes to [kratos -> processes -> translation + rotation](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/variable_redistribution_utility.h)?

**Comment by rubenzorrilla** [comment]
Now everything has more sense to me. Let me try to answer the points one by one.

> > Several concerns. Main ones:
> > 
> > * I think that the same behavior can be achieved with way less complexity (assuming that the intention is to compute both the drag and the moment).
> > * Changes in the `atomic_utilities.h` must be discussed in a standalone PR. If I'm not wrong, the method was intentionally overloaded for `array_1d<double,3>` type to ease its usage without specifying the type. Aside of this, this is the type used in Kratos vector variables.
> 
> @rubenzorrilla
> 
> There are multiple aspects:
> 
>     * the first would be to extend the drag calculation for moments as well, while keeping the existing function call possible
> 

For this I'd simply move and optimize what we did in the `ExaquteApplication`. Note that as long as you need the moment we will need to change the function call as the reference point needs to be provided. What we can discuss is if we want to overload the existent method with an extra argument or to add a new `CalculateDragAndMoment`.

>     * additional would be to remove duplicate or similar functionalities (see [ExaquteSandboxApplication -> custom_utilities -> drag_and_moment_utilities](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/ExaquteSandboxApplication/custom_utilities/drag_and_moment_utilities.h)
> 

Agree. Once this is ready we can remove the utilities in there (in a different PR).

>     * additional aspect would be to make the formulation more general as a sort of "getter" function, such that other applications can benefit from it (such as the StructuramMechanicsApplication -> to get REACTION and MOMENT_REACTION reduced to a point)
> 

If you want to get the drag or moment with respect to a point from other application there is no need to introduce any change in the `DragUtilities`. The issue would be that you will introduce a Python dependency on the `FluidDynamicsApplication` in your application of interest.

On the contrary, if your final goal is to apply the current functions to any variable, this is in my opinion out of the scope of the drag utilities and should be done in a different class.

>     * if we have a getter, one could have the inverse opretation (for example [MeshMovingApplication -> custom_utilities -> move_mesh_utilities](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/ExaquteSandboxApplication/custom_utilities/drag_and_moment_utilities.h) the could benefit from it for setting a rigid-body type translation + rotation of a model part, useful for FSI)
> 

I'm a bit confused about this... Just to be sure, you aim at having a sort of inverse operation to move mesh which would be get drag or moment? Then you will apply this to your structure in order to compute a rigid body movement right?

I like the idea, but I think that this should be done at the solver level. For me this would be a new `partitioned_rigid_body_fsi_solver.py`. The idea is that the `DragUtilities` only compute the drag (and drag moment) regardless the final use of these magnitudes.

> 
> The first two of those strictly related to functionalities of the FluidDynamicsApplication. It is a question whether already for this PR it is worth to work even on a more generic fomulation to accomocate similiar later formulations. Or do you think that we should rather think on the application level?
>

I think that it is more or less clear after the comments above, but I'd think on the application level.

> As far as making the formulation generic and potentially enabling other applications to benefit from it, should we think about something in line with what can be seen in [kratos -> utilities -> variable_redistribution_utility.h](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/variable_redistribution_utility.h)? Also make additions/changes to [kratos -> processes -> translation + rotation](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/processes/rotation_operation.h)?

You mean moving the `DragUtilities` to the core and make them general?

**Comment by mpentek** [comment]
> For this I'd simply move and optimize what we did in the `ExaquteApplication`. Note that as long as you need the moment we will need to change the function call as the reference point needs to be provided. What we can discuss is if we want to overload the existent method with an extra argument or to add a new `CalculateDragAndMoment`

Ok, so the first step will be to decide whether to overload the method or have a new implementation with a function? What would be your preference @rubenzorrilla? I guess experienced users would be the ones using it, so I could even imagine going for a new function, unless we want to avoid some code duplication. The function overload would kind of bring the benefit of less code and being able to use the original definition.

**Comment by mpentek** [comment]
> You mean moving the `DragUtilities` to the core and make them general?

Ok, here I think I need to clarify more:

Reducing reaction to a point and also calculating the moment is in an abstract view (and in lack of a better naming) a geometry-weigheted reduction, so one could see the function call as `ReduceVariableGeometryWeighed(VARIABLE_NAME, MODELPART, REFERENCE_POINT)`. It would be similar to the concept of `SumHistoricalVariable(VARIABLE, MODELPART)`.

Would other application benefit from something similar, specifically dealing with `REACTION`s? FluidDynamics and StructuralMechanics definietily would. I cannot think about another variable, othar than `REACTION`, to carry out this operation on, but perhaps there is something similar for other applications. If there would be, it could be worth to make a more generic fomulation and not to have it coded (and with code duplications) in many applications.

If we are thinking of generalizing, one could think about an inverse operation, something like `DistributeVariableGeometryWeighed(VARIABLE_NAME, MODELPART, REFERENCE_POINT)`. Here for my the clear usage would be to apply a translation and rotation of a point on a model part with the MeshMovingApp. Of course one of the use case would be an FSI with a rigid body model or simply a CFD with forced mesh movement. Other use case could be some optimization where the mesh nodes need to be moved in a similar way, the design parameter being a translation and a rotation. Again, the question is, whether such a generalization is worh doing and where to place the implementation.

**Comment by rubenzorrilla** [comment]
> > For this I'd simply move and optimize what we did in the `ExaquteApplication`. Note that as long as you need the moment we will need to change the function call as the reference point needs to be provided. What we can discuss is if we want to overload the existent method with an extra argument or to add a new `CalculateDragAndMoment`
> 
> Ok, so the first step will be to decide whether to overload the method or have a new implementation with a function? What would be your preference @rubenzorrilla? I guess experienced users would be the ones using it, so I could even imagine going for a new function, unless we want to avoid some code duplication. The function overload would kind of bring the benefit of less code and being able to use the original definition.

Both options would be technically possible. However, I think that creating a new method would be more clear for the user (what we did in the Exaqute application), specially taking into account that we need to provide the reference point for the torque. About the code duplication, I can't imagine how we can avoid this efficiently as we will always need to calculate the drag.

File: kratos/utilities/atomic_utilities.h [hunk 1]
```diff
@@ -50,11 +50,12 @@ inline void AtomicAdd(TDataType& target, const TDataType& value)
...
-inline void AtomicAdd(array_1d<double,3>& target, const array_1d<double,3>& value)
+template <std::size_t ArraySize>
+inline void AtomicAdd(array_1d<double,ArraySize>& target, const array_1d<double,ArraySize>& value)
 {
-    AtomicAdd(target[0], value[0]);
-    AtomicAdd(target[1], value[1]);
-    AtomicAdd(target[2], value[2]);
+    for (std::size_t i=0; i<ArraySize; ++i) {
+        AtomicAdd(target[i], value[i]);
+    }
```
Comment (philbucher) [inline kratos/utilities/atomic_utilities.h]: Interesting, I didn't think of this ...
Could you please make a separate PR for this? I would like to play around with it

**Comment by mpentek** [comment]
> Hence, what we can do is to add an extra `CalculateTotalForce` and `CalculateTotalForceAndTorque` to the variable utils. The first would basically call the `SumHistoricalVariables` but without a type template argument, which will always be `array_1d<double,3>`. The second one would have an extended API to provide the reference torque point as well as the output array.

I like the idea of `CalculateTotalForceAndTorque`. So should we leave the `ComputeDrag` as is in the FluidDynamicsApp and just have this new implementation in core, which for now could be used by the FluidDynamicsApp and StructuralMechanicsApp (the latter needs in addition to REACTION the REACTION_MOMENT)? Or we stick to the application-level, as it is used by 2 application only. The MeshMovinApp related enhancement than shall just be a utility for the specific app (although StructuralMechanicsApp could also benefit).

**Comment by philbucher** [comment]
to me it makes sense to have such a utility in the core, and leave the DragUtils as they are.
I agree abt hardcoding to `array3` bcs this is currently our only usecase

I would introduce a new utility for this, the variableUtils are quite big already and this doesn't really fit in there IMO

**Comment by rubenzorrilla** [comment]
> to me it makes sense to have such a utility in the core, and leave the DragUtils as they are.
> I agree abt hardcoding to `array3` bcs this is currently our only usecase
> 
> I would introduce a new utility for this, the variableUtils are quite big already and this doesn't really fit in there IMO

OK. We can create a `CalculateForceAndTorqueUtilities`.

**Comment by loumalouomega** [comment]
Do we still need this branch @matekelemen ?, can we remove it?

**Comment by matekelemen** [comment]
can be removed

**Pull request closed without merging**